snapcraft/internal/states/_global_state.py

```diff
@@ -14,23 +14,63 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+import os
+from typing import Dict, List, Type, TypeVar
+
 import yaml
 
 from snapcraft.internal.states._state import State
 
 
-def _global_state_constructor(loader, node):
-    parameters = loader.construct_mapping(node)
-    return GlobalState(**parameters)
+GlobalStateT = TypeVar("GlobalStateT", bound="GlobalState")
+
+
+class _GlobalStateLoader(yaml.Loader):
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
 
+        self.add_constructor(u"!GlobalState", type(self).construct_global_state)
```

Comment:
Yuck, they've made this awkward :stuck_out_tongue: .

Reply:
it could be added after class declaration too